### PR TITLE
Feature/Quality of life fixes

### PR DESCRIPTION
**Pull request by martinfalk**
Fixes
* link color in help widget on Windows
* broken links to doxygen API
* image diff in regression report

Changes proposed in this PR:
 * added visualizers for information processors (Image, Mesh, Volume)
 * further optimized includes

### CODE REVIEW

File: include/inviwo/core/io/serialization/serializebase.h [hunk 1]
```diff
@@ -171,46 +164,32 @@ class IVW_CORE_API SerializeBase {
...
+IVW_CORE_API void numericalFromStr(std::string_view value, unsigned short& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, int& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, unsigned int& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, long& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, unsigned long& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, long long& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, unsigned long long& dest);
+
 template <class T>
 decltype(auto) toStr(const T& value) {
```
Comment (ResearchDaniel) [inline include/inviwo/core/io/serialization/serializebase.h:182]: Deprecate and suggest to use fmt::to_string?

File: modules/base/include/modules/base/datavisualizer/imageinformationvisualizer.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <modules/base/basemoduledefine.h>
+#include <inviwo/core/rendering/datavisualizer.h>
+
+namespace inviwo {
+
+class IVW_MODULE_BASE_API ImageInformationVisualizer : public DataVisualizer {
```
Comment (ResearchDaniel) [inline modules/base/include/modules/base/datavisualizer/imageinformationvisualizer.h:36]: Could point documentation to corresponding processor

File: modules/base/include/modules/base/processors/imageinformation.h [hunk 3]
```diff
@@ -38,18 +38,6 @@
...
-/** \docpage{org.inviwo.ImageInformation, Image Information}
- * ![](org.inviwo.ImageInformation.png?classIdentifier=org.inviwo.ImageInformation)
- * Shows available information provided by input image including metadata.
- *
- * ### Inports
- *   * __image__   input image
- */
-
-/**
- * \class ImageInformation
```
Comment (ResearchDaniel) [inline modules/base/include/modules/base/processors/imageinformation.h:50]: Why remove everything?

File: modules/base/src/datavisualizer/imageinformationvisualizer.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+Document ImageInformationVisualizer::getDescription() const {
+    Document doc;
+    auto b = doc.append("html").append("body");
+    b.append("", "Construct an image information processor");
+    return doc;
+}
+
+std::vector<FileExtension> ImageInformationVisualizer::getSupportedFileExtensions() const {
+    auto rf = app_->getDataReaderFactory();
+    auto exts = rf->getExtensionsForType<Layer>();
```
Comment (ResearchDaniel) [inline modules/base/src/datavisualizer/imageinformationvisualizer.cpp:58]: Not image?

File: modules/base/src/properties/imageinformationproperty.cpp [hunk 5]
```diff
@@ -98,6 +108,7 @@ void ImageInformationProperty::updateForNewImage(const Image& image) {
     }();
 
     dimensions_.set(dim);
+    aspectRatio_.set(static_cast<double>(dim.x) / dim.y);
```
Comment (ResearchDaniel) [inline modules/base/src/properties/imageinformationproperty.cpp:107]: What if y is 0

**Review by ResearchDaniel — DISMISSED** [review]

File: include/inviwo/core/io/serialization/serializebase.h [hunk 6]
```diff
@@ -35,10 +35,11 @@
 
 #include <map>
 #include <string>
-#include <charconv>
 #include <array>
 #include <sstream>
 
+#include <fmt/format.h>
```
Comment (petersteneteg) [inline include/inviwo/core/io/serialization/serializebase.h]: fmt/core.h?

File: include/inviwo/core/util/volumeramutils.h [hunk 7]
```diff
@@ -30,8 +30,12 @@
 #pragma once
 
 #include <inviwo/core/common/inviwocoredefine.h>
-#include <inviwo/core/datastructures/volume/volumeram.h>
 #include <inviwo/core/common/inviwoapplication.h>
```
Comment (petersteneteg) [inline include/inviwo/core/util/volumeramutils.h:33]: remove?

File: modules/base/include/modules/base/processors/datasource.h [hunk 8]
```diff
@@ -96,7 +96,7 @@ class DataSource : public Processor {
      * @param content A content type passed into the FileProperty, usually 'volume', 'image' etc.
      * @see FileProperty
      */
-    DataSource(InviwoApplication* app = InviwoApplication::getPtr(), const std::string& file = "",
+    DataSource(InviwoApplication* app = util::getInviwoApplication(), const std::string& file = "",
```
Comment (petersteneteg) [inline modules/base/include/modules/base/processors/datasource.h]: remove include?

File: modules/base/include/modules/base/processors/imageinformation.h (hunk 3 above)
Comment (petersteneteg) [reply modules/base/include/modules/base/processors/imageinformation.h:50]: should be in the cpp now

File: modules/base/src/datavisualizer/imageinformationvisualizer.cpp (hunk 4 above)
Comment (petersteneteg) [reply modules/base/src/datavisualizer/imageinformationvisualizer.cpp:58]: generally a file on disk would correspond to a Layer not an Image

File: modules/base/src/properties/imageinformationproperty.cpp [hunk 9]
```diff
@@ -42,15 +42,25 @@ ImageInformationProperty::ImageInformationProperty(std::string_view identifier,
...
+    : CompositeProperty(
+          identifier, displayName,
+          "A CompositeProperty holding properties to show a information about an image"_help,
+          invalidationLevel, semantics)
+    , dimensions_("dimensions", "Dimensions", "Image dimensions"_help, size2_t(0),
+                  {size2_t(0), ConstraintBehavior::Immutable},
+                  {size2_t(std::numeric_limits<size_t>::max()), ConstraintBehavior::Immutable},
+                  size2_t(1), InvalidationLevel::Valid, PropertySemantics("Text"))
+    , aspectRatio_("aspectRatio", "Aspect Ratio",
+                   OrdinalPropertyState<double>{0.0, 0.0, ConstraintBehavior::Immutable,
```
Comment (petersteneteg) [inline modules/base/src/properties/imageinformationproperty.cpp]: there is no point in using the OrdinalPropertyState unless you plan to use on of the util::ordinalColor etc functions. here you might just as well just pass in all the args directly...

File: modules/base/src/properties/imageinformationproperty.cpp (hunk 5 above)
Comment (petersteneteg) [reply modules/base/src/properties/imageinformationproperty.cpp:107]: don't think that is valid.

File: modules/base/src/properties/meshinformationproperty.cpp [hunk 10]
```diff
@@ -43,7 +43,10 @@ MeshInformationProperty::MeshInformationProperty(std::string_view identifier,
                                                  std::string_view displayName,
                                                  InvalidationLevel invalidationLevel,
                                                  PropertySemantics semantics)
-    : CompositeProperty(identifier, displayName, invalidationLevel, semantics)
+    : CompositeProperty(
+          identifier, displayName,
+          "A CompositeProperty holding properties to show information about a mesh and its buffers"_help,
```
Comment (petersteneteg) [inline modules/base/src/properties/meshinformationproperty.cpp]: I don't think the use needs to know that is is a CompositeProperty. 
"Information about a mesh and its buffers" would be better.

File: src/qt/editor/helpwidget.cpp [hunk 11]
```diff
@@ -204,6 +207,7 @@ void link(T& factory, std::string_view id, Document::DocumentHandle& handle) {
     if (auto* fo = factory->getFactoryObject(id)) {
         auto& typeName = fo->getTypeName();
         std::string name{typeName.substr(0, typeName.find_first_of('<'))};
+        replaceInString(name, "class ", "");
```
Comment (petersteneteg) [inline src/qt/editor/helpwidget.cpp]: not sure if this is the right place...

Comment (petersteneteg) [reply src/qt/editor/helpwidget.cpp]: there might also be structs... not use if it should be in the base or not...

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: modules/base/include/modules/base/processors/datasource.h (hunk 8 above)
Comment (martinfalk) [reply modules/base/include/modules/base/processors/datasource.h]: not possible. Used in line 130 to access the DataReaderFactory
https://github.com/inviwo/inviwo/blob/be14babe4453b7a7765ea8c5a2881a0fb3b5bafa/modules/base/include/modules/base/processors/datasource.h#L130

File: src/qt/editor/helpwidget.cpp (hunk 11 above)
Comment (martinfalk) [reply src/qt/editor/helpwidget.cpp]: Would have used `util::demangle()` which, according to the docs, will remove `class ` but doesn't.

https://github.com/inviwo/inviwo/blob/bbef904fa5898e276ca5b13ba0f6c2413c3779bc/include/inviwo/core/util/demangle.h#L36-L41

File: include/inviwo/core/io/serialization/serializebase.h (hunk 1 above)
Comment (martinfalk) [reply include/inviwo/core/io/serialization/serializebase.h:182]: I would keep it for consistency's sake to match `fromStr()`.

File: modules/base/include/modules/base/datavisualizer/imageinformationvisualizer.h (hunk 2 above)
Comment (martinfalk) [reply modules/base/include/modules/base/datavisualizer/imageinformationvisualizer.h:36]: Cannot find such a thing in any of the visualizers. Future PR?

File: modules/base/src/properties/imageinformationproperty.cpp (hunk 5 above)
Comment (martinfalk) [reply modules/base/src/properties/imageinformationproperty.cpp:107]: Potentially dimensions of 0 are not valid, but might never be checked.

Comment (martinfalk) [reply modules/base/src/properties/imageinformationproperty.cpp:107]: On the other hand, what should be shown in the property then? Right now it is a double property.

File: include/inviwo/core/util/volumeramutils.h (hunk 7 above)
Comment (martinfalk) [reply include/inviwo/core/util/volumeramutils.h:33]: required in `forEachVoxel()`
https://github.com/inviwo/inviwo/blob/bbef904fa5898e276ca5b13ba0f6c2413c3779bc/include/inviwo/core/util/volumeramutils.h#L52-L55

File: include/inviwo/core/io/serialization/serializebase.h [hunk 12]
```diff
@@ -171,46 +164,32 @@ class IVW_CORE_API SerializeBase {
...
+IVW_CORE_API void numericalFromStr(std::string_view value, unsigned long& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, long long& dest);
+IVW_CORE_API void numericalFromStr(std::string_view value, unsigned long long& dest);
+
 template <class T>
 decltype(auto) toStr(const T& value) {
-    if constexpr (std::is_same_v<std::string, T>) {
-        return value;
-    } else if constexpr (std::is_same_v<std::string_view, T>) {
-        return std::string{value};
```
Comment (petersteneteg) [inline include/inviwo/core/io/serialization/serializebase.h]: We want to still keep these ones... no point in passing a string through fmt::format

File: include/inviwo/core/util/volumeramutils.h (hunk 7 above)
Comment (petersteneteg) [reply include/inviwo/core/util/volumeramutils.h:33]: I meant the inviwoapplication.h include

File: src/qt/editor/helpwidget.cpp (hunk 11 above)
Comment (petersteneteg) [reply src/qt/editor/helpwidget.cpp]: I think demangle should probably remove both class and struct on msvc

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: src/qt/editor/helpwidget.cpp [hunk 13]
```diff
@@ -207,7 +208,7 @@ void link(T& factory, std::string_view id, Document::DocumentHandle& handle) {
     if (auto* fo = factory->getFactoryObject(id)) {
         auto& typeName = fo->getTypeName();
         std::string name{typeName.substr(0, typeName.find_first_of('<'))};
-        replaceInString(name, "class ", "");
+        name = util::demangle(name.c_str());
```
Comment (petersteneteg) [inline src/qt/editor/helpwidget.cpp]: there is no need to call demangle here is it already done in PropertyFactoryObjectTemplate

**Review by petersteneteg — DISMISSED** [review]

**Comment by martinfalk** [comment]
Fixed image diff for the regression reports. We now enforce the difference image to be 8bit (instead of only showing the lower 8bit).

image

**Review by petersteneteg — APPROVED** [review]

**Pull request merged**